### DISCUSSION TITLE
A proper way to draw text?

### DISCUSSION THREAD

**Discussion by Nand39** [MonoGame/MonoGame#8235] [Graphics]
Hi! I am new to MonoGame and game dev in general, so I will be grateful for your understanding in case I'm doing everything wrong 🥲
I have a bunch of questions to ask about drawing text, and I will be very thankful for help.

- First of all, SpriteBatch.DrawString() requires SpriteFont, which defines the size of the text. Scaling will lead to resolution loss, so I need to create another sprite font every time I need to draw text with a different scale. Besides, I need to do it manually through the content pipeline. But this is very inconvenient, and I think that there might be another way to scale text. But I don't want to get into details of loading and rendering vector images yet.

- How to use shaders with text? For example, I'm making a UI system, and I want to cull half of my text so it won't stick out of some scrollable menus (even if it sounds too expensive and probably unreasonable, I still want to learn how to work with shaders and text). But shaders apply to all the characters separately, while for my previous culling shader I calculated menu rectange relative to the size of the given texture to get UV coordinates after which shader should discard pixels, however now I have many rectangles per Draw() for each character, so I don't know what to do. Besides, I don't know whether there is any way to draw different characters with different shaders, or at least different parameters. I would like to draw text as a single texture. I tried using render targets, which worked but was an awful idea because switching between them, clearing, etc. for every text object is very slow, and I definitely don't want my UI to consume as many resources as the rest of the game.

Drawing text should be fast and lightweight, but I don't see such solutions for the problems mentioned above. Thanks for you attention.

**Comment by KelsamGames** [comment]
https://github.com/FontStashSharp/FontStashSharp/wiki/Using-FontStashSharp-in-MonoGame-or-FNA

https://github.com/FontStashSharp/FontStashSharp

**Reply by Nand39** [reply]
Thank you very much for providing these links! I appreciate your help! Will definitely try this.

**Reply by ThomasFOG** [reply]
The default font pipeline in MonoGame is indeed inconvenient in that regard and FontStashSharp is a good library to consider for rendering texts.

In regard to changing shader parameters for each letter, you would have to draw each of them individually in separate batches. For clipping, I would add a bounding box feature to FontStashSharp and do the clipping from code. I don't think there is a library that do that out of the box.

(At some, when I have the time to, I'd like to contribute a better font renderer to MonoGame. In the mean time, FontStashSharp is a good start.)

**Reply by Nand39** [reply]
Thanks! That helped me a lot to define a way of creating UI. Then I will consider using external libs. Once again, thank you for help! I will close this discussion now.
